referenceframe/frame_system_test.go

```diff
@@ -237,3 +237,139 @@ func TestConvertTransformProtobufToFrameSystemPart(t *testing.T) {
+
+	inputs := make(map[string][]Input)
+	inputs["frame0"] = make([]Input, 0)
+	inputs["frame1"] = make([]Input, 0)
+	outMap, err := FrameSystemToPCD(fs, inputs)
+	test.That(t, err, test.ShouldBeNil)
+	//nolint:dupl
+	checkAgainst0 := []r3.Vector{
+		{-3, -4, -4},
+		{-5, -4, -4},
+		{-3, -4, -3},
+		{-3, -4, -5},
+		{-5, -4, -3},
+		{-5, -4, -5},
+		{-4, -3, -4},
+		{-4, -5, -4},
+		{-4, -3, -3},
+		{-4, -3, -5},
+		{-4, -5, -3},
+		{-4, -5, -5},
+		{-3, -3, -4},
+		{-3, -5, -4},
+		{-5, -3, -4},
+		{-5, -5, -4},
+		{-3, -3, -3},
+		{-3, -3, -5},
+		{-3, -5, -3},
+		{-3, -5, -5},
+		{-5, -3, -3},
+		{-5, -3, -5},
+		{-5, -5, -5},
+		{-5, -5, -3},
+		{-4, -4, -3},
+		{-4, -4, -5},
+	}
+	//nolint:dupl
+	checkAgainst1 := []r3.Vector{
+		{-1, -2, -2},
+		{-3, -2, -2},
+		{-1, -2, -1},
+		{-1, -2, -3},
+		{-3, -2, -1},
+		{-3, -2, -3},
+		{-2, -1, -2},
+		{-2, -3, -2},
+		{-2, -1, -1},
+		{-2, -1, -3},
+		{-2, -3, -1},
+		{-2, -3, -3},
+		{-1, -1, -2},
+		{-1, -3, -2},
+		{-3, -1, -2},
+		{-3, -3, -2},
+		{-1, -1, -1},
+		{-1, -1, -3},
+		{-1, -3, -1},
+		{-1, -3, -3},
+		{-3, -1, -1},
+		{-3, -1, -3},
+		{-3, -3, -3},
+		{-3, -3, -1},
+		{-2, -2, -1},
+		{-2, -2, -3},
+	}
+
+	for i, v := range outMap["frame0"] {
+		test.That(t, spatial.R3VectorAlmostEqual(v, checkAgainst0[i], 1e-2), test.ShouldBeTrue)
+	}
+	for i, v := range outMap["frame1"] {
+		test.That(t, spatial.R3VectorAlmostEqual(v, checkAgainst1[i], 1e-2), test.ShouldBeTrue)
+	}
+
+	// --------------------------------------------
+	// here we are testing an arm with a block attached to the end effector
+	fs = NewEmptySimpleFrameSystem("test")
+	logger := golog.NewTestLogger(t)
+	jsonData, err := os.ReadFile(rdkutils.ResolveFile("config/data/model_frame_geoms.json"))
+	test.That(t, err, test.ShouldBeNil)
+	model, err := UnmarshalModelJSON(jsonData, "")
+	test.That(t, err, test.ShouldBeNil)
+	orientConf, err := spatial.NewOrientationConfig(spatial.NewZeroOrientation())
+	test.That(t, err, test.ShouldBeNil)
+
+	lc := &LinkConfig{
+		ID:          "test",
+		Parent:      "world",
+		Translation: r3.Vector{1, 2, 3},
+		Orientation: orientConf,
+	}
+	lif, err := lc.ParseConfig()
+	test.That(t, err, test.ShouldBeNil)
+	// fully specified part
+	part := &FrameSystemPart{
+		FrameConfig: lif,
+		ModelFrame:  model,
+	}
+	armFrame, _, err := CreateFramesFromPart(part, logger)
+	test.That(t, err, test.ShouldBeNil)
+	fs.AddFrame(armFrame, fs.World())
+	blockName := "block"
+	blockPose := spatial.NewPoseFromPoint(r3.Vector{2, 2, 2})
+	blockdims := r3.Vector{20, 20, 20}
+	blockGeomCreator, err := spatial.NewBox(blockPose, blockdims, "box1")
+	test.That(t, err, test.ShouldBeNil)
+	blockFrame, err := NewStaticFrameWithGeometry(blockName, pose1, blockGeomCreator)
+	test.That(t, err, test.ShouldBeNil)
+	fs.AddFrame(blockFrame, armFrame)
+
+	inputs = make(map[string][]Input)
+	inputs["test"] = make([]Input, 6)
+	inputs["block"] = make([]Input, 0)
+	_, err = FrameSystemToPCD(fs, inputs)
```

Comment:
I am unsure how to validate the output of this call. The resulting number of points for the arm is ~80,000 and the number of points for the block is ~2,000.
I thought about having the output of the function be turned into a .pcd file and then doing a file comparison on the bytes, but even then the length of the bytes is extremely long.
looking for suggestions here.

Reply:
It's not necessarily necessary to check every single point.

If you pull the array of points, check the length of the array, and then spot-check the first point, last point, and a handful of other points at known indexes, that will give us confidence that the PCD as a whole is the same.

Another option could be to use a crypto library to take the MD5sum of the entire array.

Reply:
I chose to use the MD5sum however, the returned value `asBytes`, changes when I re-run the test.
This makes me believe that order in which the original .pcd was written in is subject to change. Hence, I do not think the other method of checking would work.
Lmk your thoughts.

Reply:
If ordering is changing (which seems likely as we're using lots of maps), then sorting the array should yield a deterministic thing to check against, for either method. See the "custom comparator" section here: https://yourbasic.org/golang/how-to-sort-in-go/

Writing and reading from a file is not the preferred method of converting a data structure to a byte array. See here for an example: https://stackoverflow.com/questions/16330490/in-go-how-can-i-convert-a-struct-to-a-byte-array

Reply:
gotcha, this is a cool solution!

Reply:
The first thing Peter mentioned is crucial functionality, custom comparators are awesome